logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java

```diff
@@ -1,15 +1,59 @@
 package org.logstash.launchers;
 
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
 
 import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.PrintStream;
 import java.io.StringReader;
+import java.lang.reflect.Field;
+import java.util.Map;
 
 import static org.junit.Assert.*;
 
 public class JvmOptionsParserTest {
 
+    @Rule
+    public TemporaryFolder temp = new TemporaryFolder();
+
+    private final PrintStream standardOut = System.out;
+    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
+
+    @Before
+    public void setUp() {
+        System.setOut(new PrintStream(outputStreamCaptor));
+    }
+
+    @After
+    public void tearDown() {
+        System.setOut(standardOut);
+    }
+
+    @Test
+    public void test_LS_JAVA_OPTS_isUsedWhenNoJvmOptionsIsAvailable() throws IOException, InterruptedException, ReflectiveOperationException {
+        updateEnv("LS_JAVA_OPTS", "-Xblabla");
+
+        JvmOptionsParser.main(new String[] {temp.toString()});
+
+        // Verify
+        final String output = outputStreamCaptor.toString();
+        assertEquals("Output MUST contains the options present in LS_JAVA_OPTS", "-Xblabla\n", output);
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    public static void updateEnv(String name, String val) throws ReflectiveOperationException {
+        Map<String, String> env = System.getenv();
+        Field field = env.getClass().getDeclaredField("m");
```

Comment:
I am wary of tests that reach into JVM internals to mutate them.

While I would prefer to solve this using dependency injection, perhaps we can limit the scope of changes by ensuring we "reset" state in our teardown script similar to how we do with stdout:

~~~ diff
diff --git a/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java b/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
index fd6733a7a..70447beb2 100644
--- a/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
+++ b/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
@@ -24,6 +24,8 @@ public class JvmOptionsParserTest {
     private final PrintStream standardOut = System.out;
     private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
 
+    private final String originalLsJavaOptsValue = System.getenv("LS_JAVA_OPTS");
+
     @Before
     public void setUp() {
         System.setOut(new PrintStream(outputStreamCaptor));
@@ -32,10 +34,11 @@ public class JvmOptionsParserTest {
     @After
     public void tearDown() {
         System.setOut(standardOut);
+        updateEnv("LS_JAVA_OPTS", originalLsJavaOptsValue);
     }
 
     @Test
-    public void test_LS_JAVA_OPTS_isUsedWhenNoJvmOptionsIsAvailable() throws IOException, InterruptedException, ReflectiveOperationException {
+    public void test_LS_JAVA_OPTS_isUsedWhenNoJvmOptionsIsAvailable() throws IOException, InterruptedException {
         updateEnv("LS_JAVA_OPTS", "-Xblabla");
 
         JvmOptionsParser.main(new String[] {temp.toString()});
@@ -46,11 +49,20 @@ public class JvmOptionsParserTest {
     }
 
     @SuppressWarnings({ "unchecked" })
-    public static void updateEnv(String name, String val) throws ReflectiveOperationException {
-        Map<String, String> env = System.getenv();
-        Field field = env.getClass().getDeclaredField("m");
-        field.setAccessible(true);
-        ((Map<String, String>) field.get(env)).put(name, val);
+    public static void updateEnv(String name, String val) {
+        try {
+            Map<String, String> env = System.getenv();
+            Field field = env.getClass().getDeclaredField("m");
+            field.setAccessible(true);
+            final Map<String, String> envMap = (Map<String, String>) field.get(env);
+            if (val == null) {
+                envMap.remove(name);
+            } else {
+                envMap.put(name, val);
+            }
+        } catch (ReflectiveOperationException roe) {
+            throw new RuntimeException(roe);
+        }
     }
~~~

Reply:
I agree with you about dependency injection but in this case the test calls a `main` method so we can't change its signature, unless we put an intermediate method that could accept the `JVM_OPTIONS`'s value. So in this case the `main` method simply read `JVM_OPTIONS` from environment and pass to a new method (for example named `handleJvmOptions`) which accepts application's command line arguments and environment variable value; then in the test we use this second one:

```diff
diff --git a/logstash-core/src/main/java/org/logstash/launchers/JvmOptionsParser.java b/logstash-core/src/main/java/org/logstash/launchers/JvmOptionsParser.java
index ed2d88c74..653acb9c6 100644
--- a/logstash-core/src/main/java/org/logstash/launchers/JvmOptionsParser.java
+++ b/logstash-core/src/main/java/org/logstash/launchers/JvmOptionsParser.java
@@ -68,12 +68,16 @@ public class JvmOptionsParser {
                     "Expected two arguments specifying path to LOGSTASH_HOME and an optional LS_JVM_OPTS, but was " + Arrays.toString(args)
             );
         }
+        final String lsJavaOpts = System.getenv("LS_JAVA_OPTS");
+        handleJvmOptions(args, lsJavaOpts);
+    }
 
+    static void handleJvmOptions(String[] args, String lsJavaOpts) {
         final JvmOptionsParser parser = new JvmOptionsParser(args[0]);
         final String jvmOpts = args.length == 2 ? args[1] : null;
         try {
             Optional<Path> jvmOptions = parser.lookupJvmOptionsFile(jvmOpts);
-            parser.parseAndInjectEnvironment(jvmOptions);
+            parser.parseAndInjectEnvironment(jvmOptions, lsJavaOpts);
         } catch (JvmOptionsFileParserException pex) {
             System.err.printf(Locale.ROOT,
                     "encountered [%d] error%s parsing [%s]",
@@ -108,10 +112,9 @@ public class JvmOptionsParser {
                 .findFirst();
     }
 
-    private void parseAndInjectEnvironment(Optional<Path> jvmOptionsFile) throws IOException, JvmOptionsFileParserException {
+    private void parseAndInjectEnvironment(Optional<Path> jvmOptionsFile, String lsJavaOpts) throws IOException, JvmOptionsFileParserException {
         final List<String> jvmOptionsContent = new ArrayList<>(parseJvmOptions(jvmOptionsFile));
 
-        final String lsJavaOpts = System.getenv("LS_JAVA_OPTS");
         if (lsJavaOpts != null && !lsJavaOpts.isEmpty()) {
             if (isDebugEnabled()) {
                 System.err.println("Appending jvm options from environment LS_JAVA_OPTS");
diff --git a/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java b/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
index fd6733a7a..54059e15f 100644
--- a/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
+++ b/logstash-core/src/test/java/org/logstash/launchers/JvmOptionsParserTest.java
@@ -36,9 +36,7 @@ public class JvmOptionsParserTest {
 
     @Test
     public void test_LS_JAVA_OPTS_isUsedWhenNoJvmOptionsIsAvailable() throws IOException, InterruptedException, ReflectiveOperationException {
-        updateEnv("LS_JAVA_OPTS", "-Xblabla");
-
-        JvmOptionsParser.main(new String[] {temp.toString()});
+        JvmOptionsParser.handleJvmOptions(new String[] {temp.toString()}, "-Xblabla");
 
         // Verify
         final String output = outputStreamCaptor.toString();
```

Reply:
I've taken this patch and applied it in b8a8a5c. 

Since this was a collaborative PR that you have already approved, I am going to approve it once it goes green.